### ISSUE TITLE
My proxy company adds extra layer of headers on top of the request. Is there any way to access those?

### ISSUE THREAD

**Issue by NikosEfthias** [sagebind/isahc#146]
![image](https://user-images.githubusercontent.com/9097437/71652251-94beab00-2d3d-11ea-952a-2213c86d531f.png)
In php I can grab the headers without doing anything extra but in isahc those headers simply do not exist

**Comment by sagebind** [comment]
Request headers or response headers? I'm not sure what you mean. If an outgoing proxy is adding request headers, it is impossible for you to see those, because the request has already been sent. If an incoming proxy is adding response headers then those should show up in `response.headers()` just fine, since it should be indistinguishable from headers from the actual server.

I'm not sure how PHP is relevant here, unless you mean that a PHP _server_ that is _receiving_ a request through the proxy can see the headers. Of course it can, it is receiving the request. Isahc is _sending_ the request before the proxy is involved.

Perhaps I misunderstand what you are asking.

**Comment by NikosEfthias** [comment]
sorry I meant response headers @sagebind

**Comment by NikosEfthias** [comment]
There are 2 layers of response headers as you may see in the photo 
In isahc `req.headers()` only return the latter however in php i can access both responses headers

**Comment by sagebind** [comment]
I'm not sure what you mean by "2 layers". Is the screenshot the raw response from the server? That would be a malformed HTTP/1.1 response which we make no guarantees about supporting.

Does it work with curl?

**Comment by NikosEfthias** [comment]
yes the screenshot is from curl and again with PHP curl, I can access those headers
I mean the first partial headers which I guess is not valid HTTP but still, PHP curl supports those 
But apparently they are thrown away by isahc.

**Comment by sagebind** [comment]
I'm really sorry, but I'm still not understanding what response bytes you are receiving and how they are supposed to be interpreted. Is there a redirect going on? Is the "real" HTTP response stream being enclosed as the _body_ of the outer response? I'm feeling a bit lost. I'm also confused as to how this works in PHP at all if it is a non-standard response stream.

**Comment by sagebind** [comment]
**Edit:** I see what's going on here; it looks like an HTTP CONNECT tunnel proxy is being used, and you are interested in retrieving the headers returned by the tunnel server.

> But apparently they are thrown away by isahc.

They aren't "thrown away" in that sense, they are just ignored. Isahc's goal is to return an `http::Response` that faithfully represents the response returned from the server. Merging the tunnel proxy's headers and the server's response headers into a single map would be misleading and could potentially cause issues if there are clashes in header names.

I could see adding a new API to access these, something like:

```
let response = isahc::get("https://example.org")?;

// Headers from server
for (name, value) in response.headers().iter() {
    println!("proxy: {} = {}", name, value);
}

// Headers from proxy
for (name, value) in response.proxy_headers().iter() {
    println!("proxy: {} = {}", name, value);
}
```

**Comment by NikosEfthias** [comment]
That design looks really nice

**Comment by sagebind** [comment]
Updated issue description with the proposed API and implementation notes on how this could be added.
